data-plane/src/acme/cert.rs

```diff
@@ -0,0 +1,340 @@
+            i += 1;
+
+            if i >= max_checks {
+                return Err(AcmeError::General(
+                    "Max retries for getting certificate reached".into(),
+                ));
+            }
+
+            match RawAcmeCertificate::from_storage(self.config_client.clone()).await? {
+                Some(raw_acme_certificate) => {
+                    println!("[ACME] Certificate found in storage");
+                    //Certificate already exists, decrypt it
+                    let decrypted_certificate =
+                        Self::decrypt_certificate(self.e3_client.clone(), raw_acme_certificate)
+                            .await?;
+                    persisted_certificate =
+                        Some(decrypted_certificate.to_certified_key(key.clone())?);
+                }
+                None => {
+                    println!("[ACME] Certificate not found in storage, checking for lock");
+                    let existing_lock =
+                        StorageLock::read_from_storage(CERTIFICATE_LOCK_NAME.into()).await?;
+                    match existing_lock {
+                        Some(lock) => {
+                            println!("[ACME] Lock found, checking if expired");
+                            if lock.is_expired() {
+                                println!("[ACME] Lock expired, creating new lock, creating Certificate, encrypting Certificate, persisting Certificate, deleting lock");
+                                //Lock is expired - create new lock, create Certificate - encrypt Certificate - persist Certificate - delete lock
+                                let unencrypted_certificate_maybe = self
+                                    .create_certificate_and_persist_with_lock(key.clone())
+                                    .await?;
+                                persisted_certificate = unencrypted_certificate_maybe;
+                            }
+                            println!("[ACME] Lock not expired, waiting for Certificate to be created by other instance or waiting for lock to expire");
+                            //Do nothing if Lock is not expired
+                            //wait for certificate to be created by other instance or wait for lock to expire
+                        }
+                        None => {
+                            println!("[ACME] Lock not found, creating lock, creating Certificate, encrypting Certificate, persisting Certificate, deleting lock");
+                            //Lock doesn't exist - create lock - create Certificate - encrypt Certificate - persist Certificate - delete lock
+                            let unencrypted_certificate_maybe = self
+                                .create_certificate_and_persist_with_lock(key.clone())
+                                .await?;
+                            persisted_certificate = unencrypted_certificate_maybe;
+                        }
+                    }
+                }
```

Comment:
Wdyt about splitting out the nesting here? Either into earlier returns or isolated functions? The first arm of this match (cert found in storage) doesn't need to be a match arm anyway.

Reply:
Also, any intermittent S3 error here seems to be a short-circuit/bubble up. Is that intentional?

Reply:
Yeah good shout, it was a bit of a beast. I've broken it down so the possible paths are clearer. I also added retries to the s3 calls with backoff and jitter.